File: hwy/contrib/matvec/matvec_sme_arm64.go
//go:build !noasm && darwin && arm64

package matvec

import (
	"sync"
	"unsafe"

	"github.com/ajroetker/go-highway/hwy"
)

// NOTE: SME implementation for matrix-vector multiplication.
//
// For matvec (result = M * v), we use FMOPA to process 16 rows at a time:
//   - FMOPA computes: za[i][j] += z1[i] * z2[j]
//   - Load column M[row:row+16, k] into z1
//   - Broadcast v[k] into z2
//   - After all k: za[i][0] = dot(M_row[row+i], v)
//
// To enable contiguous column access, we pre-transpose M:
//   - MT[k][i] = M[i][k] (cols×rows, row-major)
//   - Column M[row:row+16, k] becomes row MT[k, row:row+16]
//
// Apple M4 SVL = 512 bits, meaning:
//   - Z registers hold 16 × float32 or 8 × float64
//   - ZA tiles are 16×16 = 256 float32 or 8×8 = 64 float64 elements
//   - Single FMOPA does 16×16×2 = 512 FP32 ops

// Dimension thresholds for SME MatVec
// SME is faster for smaller matrices (64-192) due to 512-bit FMOPA,
// but slower for larger matrices due to transpose overhead.
// NEON's direct dot product approach is more efficient for large matrices.
const (
	minDimForSMEMatVec = 64  // Minimum dimension to use SME
	maxDimForSMEMatVec = 192 // Maximum dimension - above this NEON is faster
)

//go:noescape
func matvec_sme_f32(mt, v, result unsafe.Pointer, rows, cols int64)

//go:noescape
func matvec_sme_f64(mt, v, result unsafe.Pointer, rows, cols int64)

// Transpose buffer pools to avoid allocations
var matvecTransposePool = sync.Pool{
	New: func() interface{} {
		return make([]float32, 0, 256*256)
	},
}

var matvecTransposePool64 = sync.Pool{
	New: func() interface{} {
		return make([]float64, 0, 256*256)
	},
}

// transposeForMatVec transposes rows×cols matrix M into cols×rows matrix MT
// MT[k,i] = M[i,k]
func transposeForMatVec(m []float32, rows, cols int, mt []float32) {
	for i := 0; i < rows; i++ {
		for k := 0; k < cols; k++ {
			mt[k*rows+i] = m[i*cols+k]
		}
	}
}

// transposeForMatVec64 transposes rows×cols matrix M into cols×rows matrix MT for float64
func transposeForMatVec64(m []float64, rows, cols int, mt []float64) {
	for i := 0; i < rows; i++ {
		for k := 0; k < cols; k++ {
			mt[k*rows+i] = m[i*cols+k]
		}
	}
}

// matvecSME uses ARM SME FMOPA instruction for matrix-vector multiplication.
// Uses outer product accumulate with ZA tiles.
// Pre-transposes M for contiguous column access, enabling fast vector loads.
func matvecSME(m []float32, rows, cols int, v, result []float32) {
	// For non-aligned row sizes (16-element tiles), fall back to NEON
	if rows%16 != 0 {
		matvecNEON(m, rows, cols, v, result)
		return
	}

	// SME is faster for medium-sized matrices (64-192)
	// For smaller matrices, streaming mode overhead dominates
	// For larger matrices, transpose cost makes NEON faster
	if rows < minDimForSMEMatVec || cols < minDimForSMEMatVec ||
		rows > maxDimForSMEMatVec || cols > maxDimForSMEMatVec {
		matvecNEON(m, rows, cols, v, result)
		return
	}

	// Get transpose buffer from pool
	mtSize := rows * cols
	mtBuf := matvecTransposePool.Get().([]float32)
	if cap(mtBuf) < mtSize {
		mtBuf = make([]float32, mtSize)
	} else {
		mtBuf = mtBuf[:mtSize]
	}

	// Transpose M (rows×cols) to MT (cols×rows) for contiguous column access
	transposeForMatVec(m, rows, cols, mtBuf)

	// Call SME FMOPA with transposed M
	matvec_sme_f32(
		unsafe.Pointer(unsafe.SliceData(mtBuf)),
		unsafe.Pointer(unsafe.SliceData(v)),
		unsafe.Pointer(unsafe.SliceData(result)),
		int64(rows),
		int64(cols),
	)

	// Return buffer to pool
	matvecTransposePool.Put(mtBuf)
}

// matvecSME64 uses ARM SME FMOPA instruction for float64 matrix-vector multiplication.
// Uses 8×8 tiles for float64.
func matvecSME64(m []float64, rows, cols int, v, result []float64) {
	// For non-aligned row sizes (8-element tiles for float64), fall back to NEON
	if rows%8 != 0 {
		matvecNEON64(m, rows, cols, v, result)
		return
	}

	// SME is faster for medium-sized matrices (64-192)
	// For smaller matrices, streaming mode overhead dominates
	// For larger matrices, transpose cost makes NEON faster
	if rows < minDimForSMEMatVec || cols < minDimForSMEMatVec ||
		rows > maxDimForSMEMatVec || cols > maxDimForSMEMatVec {
		matvecNEON64(m, rows, cols, v, result)
		return
	}

	// Get transpose buffer from pool
	mtSize := rows * cols
	mtBuf := matvecTransposePool64.Get().([]float64)
	if cap(mtBuf) < mtSize {
		mtBuf = make([]float64, mtSize)
	} else {
		mtBuf = mtBuf[:mtSize]
	}

	// Transpose M (rows×cols) to MT (cols×rows) for contiguous column access
	transposeForMatVec64(m, rows, cols, mtBuf)

	// Call SME FMOPA with transposed M
	matvec_sme_f64(
		unsafe.Pointer(unsafe.SliceData(mtBuf)),
		unsafe.Pointer(unsafe.SliceData(v)),
		unsafe.Pointer(unsafe.SliceData(result)),
		int64(rows),
		int64(cols),
	)

	// Return buffer to pool
	matvecTransposePool64.Put(mtBuf)
}

func init() {
	if hwy.HasSME() {
		// Use SME FMOPA implementation for large aligned matrices
		// This overrides the generated dispatch
		MatVecFloat32 = matvecSME
		MatVecFloat64 = matvecSME64
	}
}

File: hwy/contrib/matvec/c/matvec_sme_arm64.c
// SME Matrix-Vector Multiplication for go-highway
// Compile with: -march=armv9-a+sme
//
// This implements the same algorithms as the hand-written assembly in
// hwy/contrib/matvec/matvec_sme_arm64.s
//
// The generated assembly can be compared against the hand-written version
// to verify correctness and optimize.

// GOAT's C parser uses GOAT_PARSER=1, clang doesn't
#ifndef GOAT_PARSER
#include <arm_sme.h>
#endif

// =============================================================================
// matvec_sme_f32: FMOPA-based matrix-vector multiplication
// =============================================================================
// Computes result = M * v where:
//   MT is cols x rows (M transposed, row-major) - for contiguous column access
//   v is cols (input vector)
//   result is rows (output vector)
//
// With M transposed, loading M columns becomes contiguous!
//   M[row:row+16, k] = MT[k, row:row+16] = contiguous in memory
//
// Algorithm (using outer product like matmul):
//   For each 16-row tile:
//     zero ZA
//     for k in 0..cols:
//       z2 = MT[k, row:row+16]  (contiguous load - like A column in matmul)
//       z0 = broadcast(v[k])   (like B row, but broadcast scalar)
//       FMOPA za0 += z2 ⊗ z0
//     Extract za0 rows and store first element of each to result
//
// Apple M4 SVL = 512 bits = 16 × float32
// ZA tiles = 16×16 = 256 float32 elements
//
// This mirrors: hwy/contrib/matvec/matvec_sme_arm64.s
//
// func matvec_sme_f32(mt, v, result unsafe.Pointer, rows, cols int64)
void matvec_sme_f32(float *mt, float *v, float *result,
                     long *prows, long *pcols) __arm_streaming __arm_out("za") {
    long rows = *prows;
    long cols = *pcols;

    // Process 16 rows at a time
    for (long row = 0; row < rows; row += 16) {
        // Zero accumulator tile
        svzero_za();

        // Inner loop: accumulate over columns
        for (long k = 0; k < cols; k++) {
            // Load M column: M[row:row+16, k] from transposed MT[k, row:row+16]
            // This is contiguous at: MT_base + k * rows + row
            svfloat32_t z_col = svld1_f32(svptrue_b32(), mt + k * rows + row);

            // Broadcast v[k] to all lanes
            svfloat32_t z_v = svdup_f32(v[k]);

            // Outer product accumulate: ZA0 += z_col * z_v^T
            // Since z_v is broadcast, za[i][j] = sum_k(M[row+i, k] * v[k]) for all j
            svmopa_za32_f32_m(0, svptrue_b32(), svptrue_b32(), z_col, z_v);
        }

        // Extract horizontal slices from za0 and store first element of each
        // After accumulation, za[i][j] = dot(row i, v) for all j (all columns same)
        for (int i = 0; i < 16; i++) {
            svfloat32_t zrow = svread_hor_za32_f32_m(svundef_f32(), svptrue_b32(), 0, i);
            // Extract first element and store
            result[row + i] = svlasta_f32(svptrue_b32(), zrow);
        }
    }
}

// =============================================================================
// matvec_sme_f64: FMOPA-based matrix-vector multiplication for float64
// =============================================================================
// Same algorithm but with 8-element tiles for float64.
// Apple M4 SVL = 512 bits = 8 × float64
// ZA tiles = 8×8 = 64 float64 elements
//
// This mirrors: hwy/contrib/matvec/matvec_sme_arm64.s (f64 version)
//
// func matvec_sme_f64(mt, v, result unsafe.Pointer, rows, cols int64)
void matvec_sme_f64(double *mt, double *v, double *result,
                     long *prows, long *pcols) __arm_streaming __arm_out("za") {
    long rows = *prows;
    long cols = *pcols;

    // Process 8 rows at a time (float64 has half the lanes)
    for (long row = 0; row < rows; row += 8) {
        // Zero accumulator tile
        svzero_za();

        // Inner loop: accumulate over columns
        for (long k = 0; k < cols; k++) {
            // Load M column from transposed MT[k, row:row+8]
            svfloat64_t z_col = svld1_f64(svptrue_b64(), mt + k * rows + row);

            // Broadcast v[k] to all lanes
            svfloat64_t z_v = svdup_f64(v[k]);

            // Outer product accumulate for float64
            svmopa_za64_f64_m(0, svptrue_b64(), svptrue_b64(), z_col, z_v);
        }

        // Extract and store first element of each row
        for (int i = 0; i < 8; i++) {
            svfloat64_t zrow = svread_hor_za64_f64_m(svundef_f64(), svptrue_b64(), 0, i);
            // Extract first element and store
            result[row + i] = svlasta_f64(svptrue_b64(), zrow);
        }
    }
}
